src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java

```diff
@@ -33,6 +35,26 @@ public interface BuildJobRepository extends JpaRepository<BuildJob, Long>, JpaSp
     @EntityGraph(attributePaths = { "result", "result.participation", "result.participation.exercise", "result.submission" })
     Page<BuildJob> findAll(Pageable pageable);
 
+    // Cast to string is necessary. Otherwise, the query will fail on PostgreSQL.
+    @Query("""
+            SELECT b
+            FROM BuildJob b
+                LEFT JOIN Course c ON b.courseId = c.id
+                LEFT JOIN FETCH b.result r
+                LEFT JOIN FETCH r.participation p
+                LEFT JOIN FETCH p.exercise
+                LEFT JOIN FETCH r.submission
+            WHERE (:buildStatus IS NULL OR b.buildStatus = :buildStatus)
+                AND (:buildAgentAddress IS NULL OR b.buildAgentAddress = :buildAgentAddress)
+                AND (CAST(:startDate AS string) IS NULL OR b.buildStartDate >= :startDate)
+                AND (CAST(:endDate AS string) IS NULL OR b.buildStartDate <= :endDate)
+                AND (:searchTerm IS NULL OR (b.repositoryName LIKE %:searchTerm% OR c.title LIKE %:searchTerm%))
+                AND (:courseId IS NULL OR b.courseId = :courseId)
+            """)
+    Page<BuildJob> findAllByFilterCriteria(@Param("buildStatus") BuildStatus buildStatus, @Param("buildAgentAddress") String buildAgentAddress,
+            @Param("startDate") ZonedDateTime startDate, @Param("endDate") ZonedDateTime endDate, @Param("searchTerm") String searchTerm, @Param("courseId") Long courseId,
+            Pageable pageable);
```

Comment:
Avoid using LEFT JOIN FETCH and Pageable in the same query, as this can lead to performance issues. You can refer to the following pull request for the refactoring of these queries: https://github.com/ls1intum/Artemis/pull/8741